### PR TITLE
Use pytest for the unit tests

### PR DESCRIPTION
**Pull request by nrobinaubertin**
### Overview

Use pytest for the unit tests

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: tests/unit/conftest.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+import ops
+import pytest
+from ops.model import BlockedStatus, Container
+from ops.testing import Harness
+
+from tests.unit import helpers
+from tests.unit._patched_charm import DiscourseCharm, pgsql_patch
+
+BLOCKED_STATUS = BlockedStatus.name  # type: ignore
```
Comment (jdkandersson) [inline tests/unit/conftest.py:18]: @benhoyt this shouldn't be needed anymore, right? Or is that change not released yet?

File: tests/unit/conftest.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+DATABASE_NAME = "discourse"
+
+
+@pytest.fixture(name="start_harness", autouse=True)
+def start_harness_fixture():  # -> typing.Generator[Harness, None, None]:
```
Comment (jdkandersson) [inline tests/unit/conftest.py:24]: Should the type hint either be uncommented or removed?

File: tests/unit/conftest.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    """Ops testing framework harness fixture."""
+
+    def start_harness(
```
Comment (jdkandersson) [inline tests/unit/conftest.py:27]: What is the reason for making this a fixture rather than a helper function please?

File: tests/unit/conftest.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        doesn't reinitialize the charm as expected.
+        Ref: https://github.com/canonical/operator/issues/736
+
+        Args:
+            - with_postgres: should a postgres relation be added
+            - with_redis: should a redis relation be added
+        """
+        harness = Harness(DiscourseCharm)
+        harness.disable_hooks()
+        # Protected access is needed to instantiate harness
```
Comment (jdkandersson) [inline tests/unit/conftest.py:46]: This seems to warrant a more detailed comment, I would be surprised if it is intentional to use the private harness methods to instantiate the harness

File: tests/unit/conftest.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            with helpers.patch_exec():
+                harness.container_pebble_ready("discourse")
+
+        return harness
+
+    return start_harness
+
+
+@contextlib.contextmanager
+def _patch_setup_completed():
```
Comment (jdkandersson) [inline tests/unit/conftest.py:83]: Don't we have a new way of doing this in tests with being able to interact with the file system of the container?

File: tests/unit/conftest.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    # pylint: disable=protected-access
+    harness.charm._stored.redis_relation = {
+        redis_relation_id: {"hostname": "redis-host", "port": 1010}
+    }
+
+
+def _add_ingress_relation(harness):
+    "Add ingress relation and relation data to the charm."
+    nginx_route_relation_id = harness.add_relation("nginx-route", "ingress")
+    harness.add_relation_unit(nginx_route_relation_id, "ingress/0")
```
Comment (jdkandersson) [inline tests/unit/conftest.py:134]: Shouldn't these be fixtures?

File: tests/unit/helpers.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+import contextlib
+import typing
+import unittest.mock
+
+import ops
+
+
+@contextlib.contextmanager
+def patch_exec(fail: bool = False) -> typing.Generator[unittest.mock.Mock, None, None]:
```
Comment (jdkandersson) [inline tests/unit/helpers.py:14]: Isn't there a new way of doing this now where we don't have to mock ourselves?

Comment (jdkandersson) [reply tests/unit/helpers.py:14]: Check with @weiiwang01 who helped develop it

File: tests/unit/test_charm.py [hunk 8]
```diff
@@ -6,452 +6,335 @@
...
-        """
-        arrange: given a deployed discourse charm with all the required relations
-        act: when s3_enabled configuration is True and there's no s3_bucket
-        assert: it will get to blocked status waiting for the latter.
-        """
-        self.harness.update_config(
+@pytest.mark.parametrize(
+    "with_postgres, with_redis, with_ingress, status",
+    [
+        (False, False, False, WaitingStatus("Waiting for database relation")),
```
Comment (jdkandersson) [inline tests/unit/test_charm.py:28]: Do you think it would be worth it to use `pytest.param` and use the `id` argument to describe what the case is?

File: tests/unit/test_charm.py [hunk 9]
```diff
@@ -6,452 +6,335 @@
...
+    ],
+)
+def test_relations(start_harness, with_postgres, with_redis, with_ingress, status):
+    """
+    arrange: given a deployed discourse charm
+    act: when pebble ready event is triggered
+    assert: it will have the correct status depending on the relations
+    """
+    harness = start_harness(
+        with_postgres=with_postgres, with_redis=with_redis, with_ingress=with_ingress
```
Comment (jdkandersson) [inline tests/unit/test_charm.py:45]: Do you think these should be fixtures rather than arguments to a function?

File: tests/unit/conftest.py (hunk 1 above)
Comment (benhoyt) [reply tests/unit/conftest.py:18]: I'm not quite sure which change you're referring to. However, from a quick grep, it doesn't look like this value is actually used. It's defined separately in `tests/unit/test_charm.py` and used once there as follows:

```python
    assert harness.model.unit.status.name == BLOCKED_STATUS
```

However, I think the better and more Pythonic way to write that is to use `isinstance`:

```python
    assert isinstance(harness.model.unit.status, ops.BlockedStatus)
```

Comment (jdkandersson) [reply tests/unit/conftest.py:18]: Thanks Ben

Besides not being used, the reason this `type: ignore` used to be needed is because `BlockedStatus.name` and the `.name` attribute of other statuses used to be `None | str`, I believe it was `None` in the base class. I remember a PR that changed that

File: tests/unit/conftest.py (hunk 5 above)
Comment (merkata) [reply tests/unit/conftest.py:83]: A link to an implementation would be helpful here.

Comment (jdkandersson) [reply tests/unit/conftest.py:83]: @weiiwang01 could you help please? I think this fixture uses both for the command execution and the file system: https://github.com/canonical/flask-k8s-operator/blob/609d60e723974f46269139dd2b1539e4b65fe592/tests/unit/conftest.py#L85

File: tests/unit/helpers.py (hunk 7 above)
Comment (nrobinaubertin) [reply tests/unit/helpers.py:20]: I think that you're talking about [handle_exec](https://ops.readthedocs.io/en/latest/#ops.testing.Harness.handle_exec). But `handle_exec` gives us a way to intercept a call to do something else instead, not test that the call was made.  

I tried to use `handle_exec` with a handler that would raise an Exception, like that:
```
        def test_handler(args):
            raise Exception

        harness.handle_exec(
            "discourse", 
            [],
            handler=test_handler
        )
```

But the test didn't fail. So I'm not sure that I can use `handle_exec` here.

File: tests/unit/conftest.py (hunk 2 above)
Comment (nrobinaubertin) [reply tests/unit/conftest.py]: Removed

File: tests/unit/conftest.py (hunk 3 above)
Comment (nrobinaubertin) [reply tests/unit/conftest.py]: Changed to a helper function

File: tests/unit/conftest.py (hunk 4 above)
Comment (nrobinaubertin) [reply tests/unit/conftest.py]: Removed

File: tests/unit/conftest.py (hunk 6 above)
Comment (nrobinaubertin) [reply tests/unit/conftest.py]: They're meant to modify the object created by `start_harness` (which is now a helper function). Can they be fixtures with that in mind ?

File: tests/unit/test_charm.py (hunk 8 above)
Comment (nrobinaubertin) [reply tests/unit/test_charm.py:26]: Added

File: tests/unit/test_charm.py (hunk 9 above)
Comment (nrobinaubertin) [reply tests/unit/test_charm.py:53]: I think that this is the same conversation as [here](https://github.com/canonical/discourse-k8s-operator/pull/105#discussion_r1328251271)

File: tests/unit/conftest.py (hunk 1 above)
Comment (nrobinaubertin) [reply tests/unit/conftest.py]: I rewrote this status check the way Ben proposed.

File: tests/unit/conftest.py (hunk 5 above)
Comment (nrobinaubertin) [reply tests/unit/conftest.py]: I think that you're talking about the [get_filesystem_root](https://ops.readthedocs.io/en/latest/#ops.testing.Harness.get_filesystem_root) method.  
I tried to use it that way:
```
    def _set_setup_completed():
        root = harness.get_filesystem_root("discourse")
        (root / SETUP_COMPLETED_FLAG_FILE[1:]).parent.mkdir(parents=True, exist_ok=True)
        (root / SETUP_COMPLETED_FLAG_FILE[1:]).write_text("")
    harness.charm._set_setup_completed = _set_setup_completed
```

Before realizing that the tests didn't rely on it. So I removed that.

File: tests/unit/conftest.py (hunk 6 above)
Comment (jdkandersson) [reply tests/unit/conftest.py]: Let's try this approach on here and see how we go. It is a little different to many of our charms that use fixtures to create a range of artefacts required for the test and then pick and choose which artefacts a test needs to be able to function

**Review by jdkandersson — DISMISSED** [review]

File: tests/unit/helpers.py (hunk 7 above)
Comment (jdkandersson) [reply tests/unit/helpers.py:20]: Perhaps discuss with @weiiwang01 to see if he has an idea of how to do this?

Comment (weiiwang01) [reply tests/unit/helpers.py:20]: Yes, we have talked, the reason might be that the current mocks intercepted the `ops.testing` calls.

File: tests/unit/helpers.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        exec_process_mock.wait_output = unittest.mock.MagicMock(return_value=("", ""))
+    else:
+        exec_process_mock.wait_output = unittest.mock.Mock()
+        exec_process_mock.wait_output.side_effect = ops.pebble.ExecError([], 1, "", "")
+    exec_function_mock = unittest.mock.MagicMock(return_value=exec_process_mock)
+    with unittest.mock.patch.multiple(ops.model.Container, exec=exec_function_mock):
+        yield exec_function_mock
+
+
+def start_harness(
```
Comment (amandahla) [inline tests/unit/helpers.py:39]: What an interesting approach! :clap: 

I tried to use fixtures to solve this problem of setting a harness with specific conditions for multiple tests (see discussion [here](https://github.com/canonical/synapse-operator/pull/19#discussion_r1302486670) in Synapse) but this, at least for me, it's clearer since you are using named parameters.

File: tests/unit/test_charm.py (hunk 9 above)
Comment (amandahla) [reply tests/unit/test_charm.py:53]: Just a reminder: Fixtures usage was considered hard to follow here -> https://github.com/canonical/synapse-operator/pull/19#discussion_r1302486670

**Review by amandahla — DISMISSED** [review]
LGTM

File: tests/unit/helpers.py (hunk 7 above)
Comment (nrobinaubertin) [reply tests/unit/helpers.py:20]: After discussing with Weii, we saw that I forgot to remove the patch to `Container.exec` during my test of `handle_exec`. This caused it to catch the exec call instead of `handle_exec`.  
I rewrote the tests using `handle_exec`.

**Review by jdkandersson — DISMISSED** [review]

File: tests/unit/helpers.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+    # We catch all exec calls to the container by default
+    harness.handle_exec("discourse", [], result=0)
+
+    if with_postgres:
+        _add_postgres_relation(harness)
+
+    if with_redis:
+        _add_redis_relation(harness)
+        with patch_exec():
```
Comment (weiiwang01) [inline tests/unit/helpers.py:71]: I am a little curious why the `patch_exec` is still need here since the `harness.handle_exec` has called to handle everything above.

**Review by amandahla — DISMISSED** [review]
LGTM

File: tests/unit/helpers.py (hunk 11 above)
Comment (nrobinaubertin) [reply tests/unit/helpers.py]: Forgot to remove that. Fixed

**Review by jdkandersson — APPROVED** [review]

**Review by weiiwang01 — APPROVED** [review]

**Comment by nrobinaubertin** [comment]
This PR is currently blocked by: https://github.com/canonical/discourse-k8s-operator/pull/106
